server/middleware/timeout-middleware.js

```diff
@@ -16,16 +19,29 @@ assert(requestTimeoutMs);
 // Based off of the `connect-timeout` middleware,
 // https://github.com/expressjs/timeout/blob/f2f520f335f2f2ae255d4778e908e8d38e3a4e68/index.js
 async function timeoutMiddleware(req, res, next) {
+  req.abortController = new AbortController();
+  req.abortSignal = req.abortController.signal;
+
   const timeoutId = setTimeout(() => {
+    // Signal to downstream middlewares/routes that they should stop processing/fetching
+    // things since we timed out (downstream consumers need to respect `req.abortSignal`)
+    req.abortController.abort(
+      new RouteTimeoutAbortError(
+        `Timed out after ${requestTimeoutMs}ms while trying to respond to route ${req.originalUrl}`
+      )
```

Comment:
Maybe, it would be interesting to attach some more info about the long-running requests that took up all of the time (get the spans).

Future PR stuff though.

Perhaps, it would just be better to have per-request correlated logging and see the other info that way, https://github.com/matrix-org/matrix-public-archive/issues/9